pytorch_lightning/metrics/metric.py

```diff
@@ -27,18 +39,90 @@ def __init__(self, name: str):
         """
         super().__init__()
         self.name = name
+        self._dtype = torch.get_default_dtype()
+        self._device = torch.device('cpu')
+
+        # Register hooks
+        self.register_forward_pre_hook(self.input_convert)
+        self.register_forward_hook(self.output_convert)
+        self.register_forward_hook(self.ddp_sync)
+        self.register_forward_hook(self.compute)
 
     @abstractmethod
-    def forward(self, *args, **kwargs) -> torch.Tensor:
+    def forward(self, *args, **kwargs):
         """
         Implements the actual metric computation.
 
         Returns:
-            metric value
+            metric value or metric state
 
         """
         raise NotImplementedError
 
+    def compute(self, module: nn.Module, data: Any, output: Any):
+        """
+        Implement additionally metric computations to be done after the ddp sync
+
+        Args:
+            module: current metric module
+
+            data: input to forward method
+
+            output: output from forward method
+
+        Returns:
+            final metric value
+
+        """
+        return output
+
+    def ddp_sync(self, module: nn.Module, data: Any, output: Any):
+        """
+        Implement how the outputs from forward should be synced
+
+        Args:
+            module: current metric module
+
+            data: input to forward method
+
+            output: output from forward method
+
+        Returns:
+            synced output
+
+        """
+        return output
+
+    def input_convert(self, module: nn.Module, data: Any):
+        """
+        Implement how the inputs should be casted before calling forward
+
+        Args:
+            module: current metric module
+
+            data: input to forward method
+
+        Returns:
+            casted data
+        """
+        return data
+
+    def output_convert(self, module: nn.Module, data: Any, output: Any):
+        """
+        Implement how outputs from forward should be casted
+
+        Args:
+            module: current metric module
+
+            data: input to forward method
+
+            output: output from forward method
+
+        Returns:
+            casted outputs
+        """
+        return output
+
```

Comment:
can we also add an aggregate function, that also aggregates samples on the same device?